### PR TITLE
vertical scaling support for deployment

### PR DESCRIPTION
**Pull request by pdgetrf**
1. Added vertical scaling for deployment controller
2. Refactor MergeResourceChanges to be shared between job and deployment controller
3. Moved unit tests for MergeResourceChanges to the right place
4. Added checks against QOS change in deployment

Different from job controller, originally when there's an update, deployment controller would create a new replica set hence causing pods to restart. After this PR, when there's resource change, successful or not, there will **not** be a new replicaset being created

Tested e2e on local environment.

### CODE REVIEW

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -1047,3 +1048,53 @@ func ComputeHash(template *v1.PodTemplateSpec, collisionCount *int32) uint32 {
 
 	return podTemplateSpecHasher.Sum32()
 }
+
+func MergeResourceChanges(pod *v1.Pod, cMap map[string]*v1.Container) []v1.Container {
```
Comment (vinaykul) [inline pkg/controller/controller_utils.go]: Please rename function to GetUpdatedPodResources

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -521,7 +471,7 @@ func (jm *JobController) patchJobResource(j *batch.Job, pods []*v1.Pod) error {
 			continue
 		}
```
Comment (vinaykul) [inline pkg/controller/job/job_controller.go:477]: I should have seen this earlier but please gate the changes in lines 298, 368, 563 which calls enqueueResourceUpdateForRetry in the feature gate.

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -555,6 +569,84 @@ func (dc *DeploymentController) getPodMapForDeployment(d *apps.Deployment, rsLis
...
+	for i, container := range d.Spec.Template.Spec.Containers {
+		cMap[container.Name] = &(d.Spec.Template.Spec.Containers[i])
+	}
+
+	hasResourceUpdate := false
+
+	for _, pod := range pods {
+		// Skip if a resource update is in flight
+		if _, ok := pod.Annotations[schedulerapi.AnnotationResizeResourcesRequest]; ok {
+			glog.Warningf("A resource update is in progress for pod %s. Skipping pod.", pod.Name)
```
Comment (vinaykul) [inline pkg/controller/deployment/deployment_controller.go]: Please change this to V(2).Infof per my change in the last commit.

File: pkg/controller/deployment/deployment_controller.go [hunk 4]
```diff
@@ -637,6 +734,15 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return dc.sync(d, rsList, podMap)
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.VerticalScaling) {
+		hasResourceUpdate, _ := dc.patchDeploymentResource(d)
+
+		// return early to prevent creating new replicas when doing resource update
+		if hasResourceUpdate {
```
Comment (vinaykul) [inline pkg/controller/deployment/deployment_controller.go:805]: If there are any changes in addition to resources update, we should continue.

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (pdgetrf) [reply pkg/controller/controller_utils.go]: fixed

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (pdgetrf) [reply pkg/controller/job/job_controller.go:477]: nice catch. fixed fro both job and deployment controller

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (pdgetrf) [reply pkg/controller/deployment/deployment_controller.go]: fixed

File: pkg/controller/deployment/deployment_controller.go (hunk 4 above)
Comment (pdgetrf) [reply pkg/controller/deployment/deployment_controller.go:805]: this will be fixed in subsequent PR

**Comment by vinaykul** [comment]
LGTM

**Pull request merged**